### PR TITLE
feat(train): huber loss + losses/ plugin subpackage

### PR DESCRIPTION
**Pull request by SqrtZ3**
## 背景 / 动机

`loop.py` 当前训练 loss 硬编码 `F.mse_loss(reduction="none")`,加 huber / charbonnier / 其他鲁棒 loss 都要改主循环。Flow Matching 训练里 huber + 时间相关 delta schedule 在我们小圈子的 1536 训练里收益明显(对低 t 细节端 outlier 更鲁棒,跟 detail_inv_t / min_snr 等 loss 权重正交)。

本 PR 按 ADR 0003 plugin registry 模式落地 `runtime/training/losses/` 子包,**模板完全照搬现有 [adapters/](runtime/training/adapters/) / [timestep_samplers/](runtime/training/timestep_samplers/)**:
- BUILDERS dict + `build_loss(args)` + `validate_schema_consistency()` 三件套
- `LossProtocol` 用 `runtime_checkable` Protocol(`compute(pred, target, t) -> Tensor`)
- schema 用 `loss_type: Literal["mse","huber"]` + huber 专属字段 `huber_c` / `huber_schedule`

## 改动概要

**新增 `runtime/training/losses/`(4 个文件)**:
- `protocol.py` — LossProtocol(1 必需方法 `compute`)
- `mse.py` — MseLoss,bit-for-bit 等价于历史 `F.mse_loss(reduction='none')`
- `huber.py` — HuberLoss + 三种 schedule:
  - `constant`:delta = huber_c
  - `snr`:delta = huber_c · sqrt((1-t)/t),低 t 大 delta(接近 MSE),高 t 小 delta(更鲁棒)
  - `sigma`:delta = huber_c · t/(1-t),方向相反
- `__init__.py` — BUILDERS + build_loss + validate_schema_consistency

**集成接入点(改 6 个文件)**:
- `context.py`:TrainingContext 加 `loss_fn: Any = None`
- `phases/optimizer.py`:`build_timestep_sampler` 之后挂 `ctx.loss_fn = build_loss(args)`
- `phases/bootstrap.py`:启动期校验从 3 个增到 4 个(加 `_validate_losses()`)
- `loop.py`:`loss_per_sample = F.mse_loss(...)` 换成 `ctx.loss_fn.compute(pred, target, t)`;
  InfoNoise 用的 `_raw_mse` 改为**始终单独算一次 `F.mse_loss`**,跟训练 loss 解耦
  以保证 InfoNoise 论文一致性(开销可忽略,一次 elementwise op)
- `studio/schema.py`:加 `loss_type` Literal + `huber_c` + `huber_schedule`,
  后两个 `show_when="loss_type==huber"` + advanced
- `runtime/training/README.md`:目录结构加 losses/ 描述,顶部 "5 个 plugin 子包" → "6 个",
  schema↔registry 一致性段从 "3 个 validate" → "4 个"

**测试(29 + 2)**:
- `tests/test_losses.py` 新建:29 个用例覆盖
  - MseLoss 跟 F.mse_loss bit-for-bit + 不依赖 t + Protocol 合规
  - HuberLoss quad/linear/边界 三段数学正确
  - 三种 schedule delta 的 t 依赖方向(constant 不变 / snr 低 t 大 / sigma 低 t 小)
  - plugin registry 三件套(BUILDERS keys / build_loss dispatch / validate_schema_consistency)
  - runtime_checkable Protocol
  - parametrize 4 种 loss 在 (B,C,H,W) latent shape 上输出形状 + finite + 非负
- `tests/test_plugin_registry.py` 扩展:加 `test_loss_builders_dict_has_mse_huber` 和 `test_loss_schema_consistency_passes_on_clean_dev`

## 默认行为

`loss_type` 默认 `"mse"`,所有现有 YAML / 已存 preset / 已存 version config 行为**零变化**——MseLoss.compute 跟历史 `F.mse_loss(reduction='none')` bit-for-bit 一致(单测 codify)。InfoNoise 收到的 `_raw_mse` 数学等价(改成单独算一次,不再复用 loss_per_sample,但数值相同)。

## ADR 0003 一致性

完全遵循硬规则 L113「新变体引入新 schema 字段 → subfolder」:`loss_type` 是新 Literal 枚举 + `huber_c` / `huber_schedule` 是 huber 专属字段,必须走 subfolder + BUILDERS。模板 100% 照搬 [adapters/](runtime/training/adapters/__init__.py) / [timestep_samplers/](runtime/training/timestep_samplers/__init__.py)。

注意 `loss_weighting`(min_snr / cosmap / detail_inv_t)**保持现状**(单文件多分支),没升级 plugin subfolder——它们是 *loss 权重*(Flow Matching 步级缩放系数),跟 *loss 类型*(mse / huber)正交,跟 ADR L113 描述对得上。README 在 loss_weighting.py 一行加了正交关系说明。

## 测试

```
tests/test_losses.py              29/29 PASS（新增）
tests/test_plugin_registry.py     22/22 PASS（含新增 2 条 losses 一致性测试）
tests/test_argparse_bridge.py     19/19 PASS（schema 加字段无 CLI 派生回归）
tests/test_anima_train_migration.py 8/8 PASS（YAML / CLI 契约）
tests/test_infonoise.py           16/16 PASS（关键回归：InfoNoise EMA + warmup baseline 不变）

合计 85 passed, 4 skipped, 0 failed
```

### CODE REVIEW

**Comment by WalkingMeatAxolotl** [comment]
论文出处不存在 — 作者 docstring 写"Flow Matching 下 SNR(t) = ((1-t)/t)²，sqrt(SNR) 跟 σ⁻¹ 同阶"，听起来很有道理。但：
  - SNR(t) = ((1-t)/t)² 确实是 Flow Matching 推论
  - 但 SNR 是 loss weighting 公式（出现在 min_snr / cosmap）
  - 没有任何已发表论文 / 主流 trainer（diffusers/kohya-ss/sd-scripts）用 sqrt(SNR) 作为 huber δ schedule
  - 作者在 PR 描述里说"based on FM SNR"——但 SNR 是 weighting 不是 δ schedule，这是概念混用

  EDM/Karras 2022 论文里的 huber 用法是 constant δ。"δ 随 t 变"是合理的研究方向，但需要：
  - 论文出处（哪篇 paper 提的这个公式）
  - 或上游 trainer 实现（kohya 哪个 PR 用过这种 schedule）
  - 或本地 ablation（在本项目跑过几组对比的数据）

**Comment by WalkingMeatAxolotl** [comment]
这两个参数我先隐藏了捏
